src/Core/Framework/DataAbstractionLayer/AttributeEntityCompiler.php

```diff
@@ -367,6 +374,19 @@ private function mapping(string $entity, \ReflectionProperty $property): array
             ],
         ];
 
+        // If the target entity is versioned, we need to add the [entity]_version_id field
+        if ($registry->getByClassOrEntityName($field->entity)->isVersionAware()) {
```

Comment:
i guess this will not work in any case 🙈 .. When you reference another entity, which is not already registered as an entity (also via attributes), it is not in the definition instance registry.

I suspect that we have to run afterwards through all definitions and add the reference version fields, after all entities were registered

```

$this->registerEntities();

$entities = $this->registry->getEntities();

foreach ($entities as $entity) {
   // detect foreign keys
  // if reference is version_aware, add reference version field for foreign key name
}
```

Reply:
Hey, like described, the fix only would work for mapping entities, but you also have other cases where you rely on some meta information about entities which 
a) may not already processed in the loop of the compiler
b) come from custom entities
c) rely on the "old" definition which are used currently.

Therefore i would go in a direction where we split the compiling of the entities into two steps:
1) Compile all entities via php-attributes and collect as much information as possible
2) Process all compiled entities and enrich with global information

So IMO the compiler pass `compile` function should look in a way like this:

```php

    public function process(ContainerBuilder $container): void
    {
        $services = $container->findTaggedServiceIds('shopware.entity');

        $definitions = [];
        // first loop all services and collect all definitions and entities
        foreach ($services as $class => $_) {
            /** @var class-string<object> $class */
            $x = $this->compiler->compile($class);

            foreach ($x as $definition) {
                $definitions[] = $definition;
            }
        }

        $registry = $container->get(DefinitionInstanceRegistry::class);

        // after we compiled all definitions, we can process them to enrich them with additional information
        // like information we need from other definitions which were not already processed in the first loop
        // we also provide the registry (ATM) to have access to none php-attribute entities (custom entities, <= v6.6 definitions)
        $definitions = $this->compiler->process($definitions, $registry);

        // after we processed all definitions, we can now create the services (repositories, definitions, translations, ...)
        foreach ($definitions as $definition) {
            if ($definition['type'] === 'entity') {
                $this->definition($definition, $container, $definition['entity_name']);

                $this->repository($container, $definition['entity_name']);

                $this->translation($definition, $container, $definition['entity_name']);

                continue;
            }

            if ($definition['type'] === 'mapping') {
                $this->mapping($definition, $container);
            }
        }
    }
```

Reply:
I'm a bit confused though: now that I moved it to the defineFields method, isn't this happening after everything is already loaded?

I tested in my own project and it seems to both work when I link to order (a non atttibuted entity) as well as entities with attributes that are loaded in a "different" order.

Reply:
LGTM

Reply:
I changed the code a bit, that also `source` is considered.. so we have now `reference` and `source`

Reply:
Nice thanks. I see some conflicts are there now so will try to resolve them next week

Reply:
I fixed the conflict. But I don't know what you mean with your remark that you changed the code a bit, I don't see any changes in my PR? Or do you mean the code for the attribute logic itself? Do I need any changes for that?